### PR TITLE
recipeEditor to be able to import from experimenter

### LINKED ISSUE #81: Create import recipe page
This would just be a version of the create/edit page that makes a call to experimenters API and grabs the appropriate recipe data to pre-populate the form. 

It will also populate a special "Import instructions" field that must be cleared before the user can proceed with saving the recipe.

It should live at:
`recipes/import/:experimenterSlug`

### PR DESCRIPTION
**Pull request by tiftran**
Fixes #81

### CODE REVIEW

**Comment by jaredlockhart** [comment]
I remember when I spent a summer building the first version of this to import from Experimenter into 'Normandy Admin', as it was called in the day.  Poor PR got sent off on an ice float to live alone with the polar bears, never to be seen again.  We've come so far.  Super stoked for this @tiftran, thanks! 😄

File: extension/content/utils/normandyApi.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import Api from "devtools/utils/api";
+export default class NormandyAPI extends Api {
```
Comment (rehandalal) [inline extension/content/utils/normandyApi.js]: nit: you need a newline after imports

File: extension/content/utils/experimenterApi.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import Api from "devtools/utils/api";
+export default class ExperimenterAPI extends Api {
```
Comment (rehandalal) [inline extension/content/utils/experimenterApi.js]: nit: new line

File: extension/content/utils/experimenterApi.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  getBaseUrl() {
+    return new URL("api/v1/", this.environment.experimenterUrl).href;
+  }
+  async request({ url, extraHeaders, ...options }) {
```
Comment (rehandalal) [inline extension/content/utils/experimenterApi.js]: nit: new line between functions

File: extension/content/utils/api.js [hunk 4]
```diff
@@ -1,67 +1,26 @@
...
   }
-
-  getBaseUrl({ version = 3, mode }) {
-    const base =
-      mode.toLowerCase() === "w"
-        ? this.environment.writeableUrl
-        : this.environment.readOnlyUrl;
-
-    return new URL(`api/v${version}/`, base).href;
+  getBaseUrl() {
```
Comment (rehandalal) [inline extension/content/utils/api.js:5]: nit: new line between functions

File: extension/content/utils/normandyApi.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    const base =
+      mode.toLowerCase() === "w"
+        ? this.environment.writeableUrl
+        : this.environment.readOnlyUrl;
+
+    return new URL(`api/v${version}/`, base).href;
+  }
+
+  async request({ version = 3, url, extraHeaders, ...options }) {
+    const settings = super.formatHeaders({ extraHeaders, ...options });
```
Comment (rehandalal) [inline extension/content/utils/normandyApi.js]: this should also probably be `this`?

File: extension/content/utils/api.js [hunk 6]
```diff
@@ -1,67 +1,26 @@
...
-      headers.append(headerName, extraHeaders[headerName]);
-    }
-
-    const settings = {
+    return {
       headers,
       method: "GET",
       ...options,
     };
+  }
```
Comment (rehandalal) [inline extension/content/utils/api.js]: i dont think we need this function this should continue to be in `request`

File: extension/content/utils/api.js [hunk 7]
```diff
@@ -1,67 +1,26 @@
-export default class NormandyAPI {
```
Comment (rehandalal) [inline extension/content/utils/api.js:1]: Here's what I think this file should look like:

```diff
--- a/extension/content/utils/api.js
+++ b/extension/content/utils/api.js
@@ -1,26 +1,50 @@
-export default class Api {
+export default class API {
   constructor(environment) {
     this.environment = environment;
   }
+
   getBaseUrl() {
     throw new Error("getBaseURL() needs to be implemented");
   }
 
-  formatHeaders({ extraHeaders, ...options }) {
+  async request({ url, extraHeaders, version, ...options }) {
     const headers = new Headers();
     headers.append("Accept", "application/json");
     if (!(options.body && options.body instanceof FormData)) {
       headers.append("Content-Type", "application/json");
     }
-    return {
+
+    for (let headerName in extraHeaders) {
+      headers.append(headerName, extraHeaders[headerName]);
+    }
+
+    const settings = {
       headers,
       method: "GET",
       ...options,
     };
-  }
 
-  async request(url, settings) {
-    const response = await fetch(url, settings);
+    const apiUrl = new URL(url, this.getBaseUrl({ version, ...settings }));
+
+    if ("data" in settings) {
+      if ("body" in settings) {
+        throw new Error(
+          "Only pass one of `settings.data` and `settings.body`.",
+        );
+      }
+
+      if (["GET", "HEAD"].includes(settings.method.toUpperCase())) {
+        Object.entries(settings.data).forEach(([key, value]) => {
+          apiUrl.searchParams.append(key, value);
+        });
+      } else {
+        settings.body = JSON.stringify(settings.data);
+      }
+
+      delete settings.data;
+    }
+
+    const response = await fetch(apiUrl, settings);
 
     if (!response.ok) {
       let message;
```

File: extension/content/utils/experimenterApi.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import Api from "devtools/utils/api";
```
Comment (rehandalal) [inline extension/content/utils/experimenterApi.js]: Changes here with changes to `api.js`:
```diff
--- a/extension/content/utils/experimenterApi.js
+++ b/extension/content/utils/experimenterApi.js
@@ -1,12 +1,8 @@
-import Api from "devtools/utils/api";
-export default class ExperimenterAPI extends Api {
-  getBaseUrl() {
-    return new URL("api/v1/", this.environment.experimenterUrl).href;
-  }
-  async request({ url, extraHeaders, ...options }) {
-    const apiUrl = new URL(url, this.getBaseUrl());
-    const settings = super.formatHeaders({ extraHeaders, ...options });
-    return super.request(apiUrl, settings);
+import API from "devtools/utils/api";
+
+export default class ExperimenterAPI extends API {
+  getBaseUrl({ version = 1 }) {
+    return new URL(`api/v${version}/`, this.environment.experimenterUrl).href;
   }
 
   async fetchRecipe(slug) {

```

File: extension/content/utils/normandyApi.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
+import Api from "devtools/utils/api";
```
Comment (rehandalal) [inline extension/content/utils/normandyApi.js]: And this one, after changes to `api.js`:
```diff
--- a/extension/content/utils/normandyApi.js
+++ b/extension/content/utils/normandyApi.js
@@ -1,54 +1,34 @@
-import Api from "devtools/utils/api";
-export default class NormandyAPI extends Api {
+import API from "devtools/utils/api";
+
+export default class NormandyAPI extends API {
   constructor(environment, auth) {
     super(environment);
     this.auth = auth;
   }
 
-  getBaseUrl({ version = 3, mode }) {
-    const base =
-      mode.toLowerCase() === "w"
-        ? this.environment.writeableUrl
-        : this.environment.readOnlyUrl;
-
+  getBaseUrl({ version = 3, method }) {
+    const isReadOperation = ["GET", "HEAD"].includes(method.toUpperCase());
+    const base = isReadOperation
+      ? this.environment.readOnlyUrl
+      : this.environment.writeableUrl;
     return new URL(`api/v${version}/`, base).href;
   }
 
-  async request({ version = 3, url, extraHeaders, ...options }) {
-    const settings = super.formatHeaders({ extraHeaders, ...options });
-
-    const isReadOperation = ["GET", "HEAD"].includes(
-      settings.method.toUpperCase(),
-    );
-    const mode = isReadOperation ? "r" : "w";
-    const apiUrl = new URL(url, this.getBaseUrl({ version, mode }));
+  async request(options) {
+    const { extraHeaders = [], method = "GET" } = options;
+    const isReadOperation = ["GET", "HEAD"].includes(method.toUpperCase());
 
     if (!isReadOperation) {
-      settings.headers.append(
+      extraHeaders.append(
         "Authorization",
         `Bearer  ${this.auth.result.accessToken}`,
       );
     }
 
-    if ("data" in settings) {
-      if ("body" in settings) {
-        throw new Error(
-          "Only pass one of `settings.data` and `settings.body`.",
-        );
-      }
-
-      if (isReadOperation) {
-        Object.entries(settings.data).forEach(([key, value]) => {
-          apiUrl.searchParams.append(key, value);
-        });
-      } else {
-        settings.body = JSON.stringify(settings.data);
-      }
-
-      delete settings.data;
-    }
-
-    return super.request(apiUrl, settings);
+    return super.request({
+      ...options,
+      extraHeaders,
+    });
   }
 
   fetchRecipePage(page, searchParams = {}) {
```

**Review by rehandalal — CHANGES_REQUESTED** [review]
Nice job! 

Just a little refactor needed on the API stuff.

**Comment by rehandalal** [comment]
just gotta get master merged in to this

File: extension/content/components/recipes/RecipeEditor.js [hunk 10]
```diff
@@ -131,14 +175,23 @@ export default function RecipeEditor(props) {
...
         </FormGroup>
+        <FormGroup hidden={!match.params.slug}>
+          <ControlLabel>Import Instuctions</ControlLabel>
+          <Input
+            componentClass="textarea"
+            rows={3}
+            value={importInstructions}
+            onChange={(value) => setImportInstructions(value)}
+          />
+        </FormGroup>
```
Comment (rehandalal) [inline extension/content/components/recipes/RecipeEditor.js]: I'd rather see this field disappear altogether on non-import pages than just be hidden

**Review by rehandalal — APPROVED** [review]
THANK YOU!! GREAT WORK! 👏🏽

bors r+

**Pull request merged**